protocols/imap/src/test/java/org/apache/james/imap/processor/base/UidMsnConverterTest.java

```diff
@@ -248,23 +248,87 @@ public void addUidShouldLeadToValidConversionWhenInsertInFirstPosition() {
         testee.addUid(messageUid1);
 
         assertThat(mapTesteeInternalDataToMsnByUid().entrySet())
-            .containsOnlyElementsOf(ImmutableMap.of(
+            .containsExactlyElementsOf(ImmutableMap.of(
                 1, messageUid1,
                 2, messageUid2,
                 3, messageUid3,
                 4, messageUid4).entrySet());
     }
 
     @Test
-    public void constructorWithOutOfOrderIteratorShouldLeadToValidConversion() {
-        testee = new UidMsnConverter(ImmutableList.of(messageUid2,
+    public void addAllShouldLeadToValidConversion() {
+        testee.addAll(ImmutableList.of(messageUid1,
+            messageUid2,
+            messageUid3,
+            messageUid4));
+
+        assertThat(mapTesteeInternalDataToMsnByUid().entrySet())
+            .containsExactlyElementsOf(ImmutableMap.of(
+                1, messageUid1,
+                2, messageUid2,
+                3, messageUid3,
+                4, messageUid4).entrySet());
+    }
+
+    @Test
+    public void addAllShouldDeduplicateElements() {
+        testee.addUid(messageUid1);
+
+        testee.addAll(ImmutableList.of(messageUid1,
+            messageUid2,
+            messageUid3,
+            messageUid4));
+
+        assertThat(mapTesteeInternalDataToMsnByUid().entrySet())
+            .containsExactlyElementsOf(ImmutableMap.of(
+                1, messageUid1,
+                2, messageUid2,
+                3, messageUid3,
+                4, messageUid4).entrySet());
+    }
+
+    @Test
+    public void addAllShouldMergeWithPreviousData() {
+        testee.addUid(messageUid1);
+
+        testee.addAll(ImmutableList.of(messageUid2,
+            messageUid3,
+            messageUid4));
+
+        assertThat(mapTesteeInternalDataToMsnByUid().entrySet())
+            .containsExactlyElementsOf(ImmutableMap.of(
+                1, messageUid1,
+                2, messageUid2,
+                3, messageUid3,
+                4, messageUid4).entrySet());
+    }
+
+    @Test
+    public void addAllShouldMergeAndDeduplicatePreviousData() {
+        testee.addUid(messageUid1);
+        testee.addUid(messageUid3);
+
+        testee.addAll(ImmutableList.of(messageUid2,
+            messageUid3,
+            messageUid4));
+
+        assertThat(mapTesteeInternalDataToMsnByUid().entrySet())
+            .containsExactlyElementsOf(ImmutableMap.of(
+                1, messageUid1,
+                2, messageUid2,
+                3, messageUid3,
+                4, messageUid4).entrySet());
+    }
+
+    @Test
+    public void addAllWithOutOfOrderIteratorShouldLeadToValidConversion() {
```

Comment:
It's unfair, it's used everywhere in the class:

Valid => MonoticMSNToUIDConversion would be better?